File: sections/27_bitmask_dp/problems/d_compatible_lineup/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return

    it = iter(data)
    n = next(it)
    m = next(it)
    score = [next(it) for _ in range(n)]
    incompatible = [0] * n
    for _ in range(m):
        u = next(it) - 1
        v = next(it) - 1
        incompatible[u] |= 1 << v
        incompatible[v] |= 1 << u

    total_masks = 1 << n
    total = [0] * total_masks
    compatible = bytearray(total_masks)
    compatible[0] = 1
    answer = 0

    for mask in range(1, total_masks):
        lowest_bit = mask & -mask
        performer = lowest_bit.bit_length() - 1
        previous = mask ^ lowest_bit
        total[mask] = total[previous] + score[performer]
        if compatible[previous] and not (incompatible[performer] & previous):
            compatible[mask] = 1
            if total[mask] > answer:
                answer = total[mask]

    print(answer)


if __name__ == "__main__":
    main()

File: sections/27_bitmask_dp/problems/d_compatible_lineup/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n, m;
    if (!(cin >> n >> m)) return 0;

    vector<long long> score(n);
    for (long long& value : score) cin >> value;

    vector<int> incompatible(n, 0);
    for (int i = 0; i < m; ++i) {
        int u, v;
        cin >> u >> v;
        --u;
        --v;
        incompatible[u] |= 1 << v;
        incompatible[v] |= 1 << u;
    }

    int total_masks = 1 << n;
    vector<long long> total(total_masks, 0);
    vector<char> compatible(total_masks, false);
    compatible[0] = true;
    long long answer = 0;

    for (int mask = 1; mask < total_masks; ++mask) {
        int bit = __builtin_ctz(mask);
        int previous = mask ^ (1 << bit);
        total[mask] = total[previous] + score[bit];
        compatible[mask] = compatible[previous] &&
                           ((incompatible[bit] & previous) == 0);
        if (compatible[mask]) answer = max(answer, total[mask]);
    }

    cout << answer << '\n';
    return 0;
}
